src/Compilers/CSharp/Portable/Compiler/Instrumentation.cs

```diff
@@ -0,0 +1,212 @@
+                    // recomputed in each method prologue.
+                    BoundExpression mvid = factory.Property(factory.Property(factory.TypeofBeforeRewriting(method.ContainingType), "Module"), "ModuleVersionId");
+                    BoundExpression methodToken = factory.MethodToken(method);
+                    BoundStatement createPayloadCall = factory.ExpressionStatement(factory.Call(null, createPayload, mvid, methodToken, factory.Field(null, payloadField), factory.Literal(dynamicAnalysisSpans.Length)));
+
+                    BoundExpression payloadNullTest = factory.Binary(BinaryOperatorKind.ObjectEqual, boolType, factory.Field(null, payloadField), factory.Null(payloadType));
+                    BoundStatement payloadIf = factory.If(payloadNullTest, createPayloadCall);
+
+                    ImmutableArray<BoundStatement> newStatements = newMethodBody.Statements.Insert(0, payloadIf);
+                    newMethodBody = newMethodBody.Update(newMethodBody.Locals, newMethodBody.LocalFunctions, newStatements);
+
+                    if (IsTestMethod(method))
+                    {
+                        // If the method is a test method, wrap the body in:
+                        //
+                        // Instrumentation.FlushPayload();
+                        // try
+                        // {
+                        //     ... body ...
+                        // }
+                        // finally
+                        // {
+                        //     Instrumentation.FlushPayload();
+                        // }
+
+                        BoundStatement firstFlush = factory.ExpressionStatement(factory.Call(null, flushPayload));
+                        BoundStatement secondFlush = factory.ExpressionStatement(factory.Call(null, flushPayload));
+                        BoundStatement tryFinally = factory.Try(newMethodBody, ImmutableArray<BoundCatchBlock>.Empty, factory.Block(ImmutableArray.Create(secondFlush)));
+                        newMethodBody = factory.Block(ImmutableArray.Create(firstFlush, tryFinally));
+                    }
+
+                    return newMethodBody;
+                }
+            }
+
+            dynamicAnalysisSpans = ImmutableArray<SourceSpan>.Empty;
+            return methodBody;
+        }
+
+        private static FieldSymbol GetPayloadField(MethodSymbol method, int methodOrdinal, TypeSymbol payloadType, SyntheticBoundNodeFactory factory, CSharpCompilation compilation)
+        {
+            // ToDo PROTOTYPE (https://github.com/dotnet/roslyn/issues/9810):
+            // If the type containing the method is generic, synthesize a helper type and put the payload field there.
+            // If the payload field is part of a generic type, there will be a new instance of the field per instantiation of the generic,
+            // and so the payload field must be a member of another type.
```

Comment:
Why would someone want to reflect over instrumented types? Even if they do, I don't think they should expect to find things as defined in user code because it is by definition, instrumented. I am not against putting these fields in another type but am just curious if there is some scenario that I am missing?